practice/ttt/board.clj

```diff
@@ -0,0 +1,84 @@
+(ns ttt.board
+  (:refer-clojure :exclude [update])
+  (:require
+    [clojure.set]
+    [ttt.matrix-operations :as mo]))
+
+
+(def size 3)
+
+
+(def position-to-coordinate
+  (->> (for [x (range size)
+             y (range size)]
+         [x y])
+       (map-indexed (fn [i coord] [(inc i) coord]))
+       (into {})))
+
+
+(def empty-board
+  (mapv vec (partition size (repeat (* size size) :e))))
+
+
+(def valid-positions
+  (keys position-to-coordinate))
+
+
+(def positions-string
+  (mo/matrix->string (partition size (sort valid-positions))))
+
+
+(def valid-game-pieces-set #{:o :x :e})
+
+(def game-pieces-set #{:o :x})
+
+
+(defn to-string
+  [board]
+  (mo/matrix->string (mapv #(mapv name %) board)))
+
+
+(defn empty-coordinate?
+  [board coordinate]
+  (= :e (get-in board coordinate)))
+
+
+(defn valid?
+  [board]
+  (and (mo/square-matrix? board)
+       (clojure.set/subset? (set (flatten board)) valid-game-pieces-set)))
+
+
+(defn update
+  [board coordinate game-piece]
+  (when (valid? board)
+    (assoc-in board coordinate game-piece)))
+
+
+(defn winner-of-collection
```

Comment:
question: What does `collection` refer to?

Reply:
the function can find the winner of any collection as opposed to a 3x3 board. Used the word collection to generalize